Logs Tab: Show log line count

gtk-grafana opened the pull request:
Fixes: https://github.com/grafana/explore-logs/issues/425

<img width="1728" alt="image" src="https://github.com/user-attachments/assets/d9623be2-d510-4952-aba0-7f0645a9c27d">

<img width="594" alt="image" src="https://github.com/user-attachments/assets/5a67936b-15a4-43a7-8185-93eefce792a7">

<img width="1721" alt="image" src="https://github.com/user-attachments/assets/fb5dd56b-05bf-4c68-8724-d7e2fa426bb5">

<img width="1719" alt="image" src="https://github.com/user-attachments/assets/2f78ef34-a152-45e4-933a-6ff67646d3ea">

\@TODO:
* ✅  we're running this query on every activation of the ServiceScene, can we only run it when the query/timerange changes or if the current count is undefined?
* ✅  fix e2e assertions now that there's another query in the mix
* ✅  LogsCount needs to sum all infinite scroll pages
* ✅ Custom maxLines support

gtk-grafana commented on src/Components/ServiceScene/ActionBarScene.tsx:
```diff
@@ -168,3 +175,55 @@ function ToolbarExtensionsRenderer(props: { serviceScene: SceneObject }) {
+
+    const formattedLogsCount = valueFormatter(logsCount, 0);
+
+    if (logsCount < LINE_LIMIT || totalCount === undefined) {
```
Hmm, will this get messed up by infinite scrolling?

gtk-grafana replied:
Not this, but the logsCount total does not sum multiple pages

gtk-grafana commented:
@matyax There doesn't seem to be any way for the app to know when a new page has been loaded in the LogsPanel with infinite scroll, the observable is not updated, so the panel never goes into the loading state, and there's nothing for us to subscribe to from the app.

Do we need to emit an event from the panel plugin?

matyax commented:
Would maybe this help? https://github.com/grafana/explore-logs/blob/main/src/Components/ServiceScene/LogsPanelScene.tsx#L128

svennergr commented on src/services/query.ts:
```diff
@@ -225,3 +228,5 @@ export function unwrapWildcardSearch(input: string) {
   return expression.replace(/\s*,\s*}/, '}');
 }
+
+export const LINE_LIMIT = 1000;
```
```suggestion
// default line limit; each data source can define it's own line limit too
export const LINE_LIMIT = 1000;
```

svennergr commented on src/Components/ServiceScene/ServiceScene.tsx:
```diff
@@ -397,8 +430,29 @@ export class ServiceScene extends SceneObjectBase<ServiceSceneState> {
+    return this.state.$data?.subscribeToState((newState, prevState) => {
       this.updateLoadingState(newState, TabNames.logs);
+      if (newState.data?.state === LoadingState.Done || newState.data?.state === LoadingState.Streaming) {
+        const resultCount = newState.data.series[0].length;
```
```suggestion
        const resultCount = newState.data.series[0]?.length ?? 0;
```
Just as a safeguard if there's no series?

svennergr commented on src/Components/ServiceScene/ServiceScene.tsx:
```diff
@@ -397,8 +430,29 @@ export class ServiceScene extends SceneObjectBase<ServiceSceneState> {
+    return this.state.$logsCount?.subscribeToState((newState) => {
+      if (newState.data?.state === LoadingState.Done) {
+        const value: number | undefined = newState.data.series[0]?.fields?.[1]?.values?.[0];
+        if (value !== undefined) {
```
What happens if it is undefined and the `totalLogsCount` state is currently set? Should we reset to 0?

gtk-grafana replied:
If the instant query returns "Done" but does not return a result, I guess that would have to be a no-data state, so yeah we should set 0, but the total count won't get displayed if the logs query returns 0, so hopefully it doesn't matter either way.

Updated!

svennergr reviewed (approved):
Left a few nits. For my own taste, we could spare the displayed log count and just show the number of total logs, but that's on you and @zizzpudding, I'd say.

gtk-grafana commented:
@svennergr I chatted with @zizzpudding, we decided to move the log line count to the logs/table panel header, and additionally place the total log line count in the log volume panel header.

gtk-grafana commented:
@svennergr I updated the UI, however in my testing I discovered a routing bug that was causing labels containing `/` to redirect back to the same route on every label update, but more importantly I realized how "off" the instant query results are compared to the log volume query, particularly for low volume/ small duration queries.

<img width="1728" alt="image" src="https://github.com/user-attachments/assets/c8118982-ffca-4cd6-bc9a-0ad7039a140d" />

So I added the check back in there to output the results of the logs panel query if the instant query is undefined, and the logs panel returned less then the line count. 

<img width="1728" alt="image" src="https://github.com/user-attachments/assets/e54dcb1e-3d89-4493-99cb-92c5119d2a6d" />

gtk-grafana commented:
Oops we forgot the table header

gtk-grafana commented:
@svennergr I added the count to the table as well.
@matyax I found a bug with the data frame: https://github.com/grafana/explore-logs/issues/966

Also we might want to remove the logs count in the footer now, but let's do that in another PR.
